File: server/app/routers/google_data.py
import json
from pathlib import Path
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from server.app.config import Config
from server.app.database import get_db
from server.app.routers.settings import require_auth
from server.app.services.encryption import load_or_create_key
from server.app.services.google_auth import get_valid_access_token
from server.app.services.google_calendar import fetch_calendars
from server.app.services.google_photos import (
    create_picker_session,
    get_picker_session,
    get_session_media_items,
    delete_picker_session,
)

router = APIRouter(prefix="/api/google", tags=["google_data"], dependencies=[Depends(require_auth)])

# Module-level config; set via set_config() during app startup or tests.
_config: Optional[Config] = None


def set_config(config: Config) -> None:
    global _config
    _config = config


def _key_path() -> Path:
    assert _config is not None, "Config not set; call set_config() first"
    return _config.secret_key_path


def _settings_path() -> Path:
    assert _config is not None, "Config not set; call set_config() first"
    return _config.db_path.parent / "settings.json"


async def _get_access_token(db: Session) -> str:
    key = load_or_create_key(_key_path())
    settings_file = _settings_path()
    settings = {}
    if settings_file.exists():
        settings = json.loads(settings_file.read_text())
    client_id = settings.get("google_client_id", "")
    client_secret = settings.get("google_client_secret", "")

    token = await get_valid_access_token(
        session=db,
        encryption_key=key,
        client_id=client_id,
        client_secret=client_secret,
    )
    if not token:
        raise HTTPException(status_code=400, detail="Google not connected")
    return token


@router.get("/calendars")
async def get_calendars(db: Session = Depends(get_db)):
    access_token = await _get_access_token(db)
    return await fetch_calendars(access_token=access_token)


@router.post("/photos/picker-session")
async def create_photo_picker_session(db: Session = Depends(get_db)):
    access_token = await _get_access_token(db)
    session_data = await create_picker_session(access_token=access_token)
    return {
        "session_id": session_data["id"],
        "picker_uri": session_data["pickerUri"],
        "polling_config": session_data.get("pollingConfig", {}),
    }


@router.get("/photos/picker-session/{session_id}")
async def poll_photo_picker_session(session_id: str, db: Session = Depends(get_db)):
    access_token = await _get_access_token(db)
    session_data = await get_picker_session(access_token=access_token, session_id=session_id)
    media_items_set = session_data.get("mediaItemsSet", False)
    result: dict = {"media_items_set": media_items_set}
    if media_items_set:
        from urllib.parse import quote
        items = await get_session_media_items(access_token=access_token, session_id=session_id)
        result["photos"] = [
            {
                "id": item["id"],
                "url": f"/api/photos/proxy?url={quote(item['baseUrl'], safe='')}",
                "mimeType": item["mimeType"],
            }
            for item in items
        ]
    return result


@router.delete("/photos/picker-session/{session_id}")
async def delete_photo_picker_session(session_id: str, db: Session = Depends(get_db)):
    access_token = await _get_access_token(db)
    await delete_picker_session(access_token=access_token, session_id=session_id)

File: src/server/routes/google-data.ts
import { readFileSync, existsSync } from 'fs'
import { dirname, join } from 'path'
import type { FastifyInstance } from 'fastify'
import { requireAuth } from '../middleware/auth.js'
import { loadOrCreateKey } from '../services/encryption.js'
import { getValidAccessToken } from '../services/google-auth.js'
import { fetchCalendars } from '../services/google-calendar.js'
import {
  createPickerSession,
  getPickerSession,
  getSessionMediaItems,
  deletePickerSession,
} from '../services/google-photos.js'

function loadSettings(config: { dbPath: string }): Record<string, unknown> {
  const path = join(dirname(config.dbPath), 'settings.json')
  if (existsSync(path)) {
    return JSON.parse(readFileSync(path, 'utf-8'))
  }
  return {}
}

async function getAccessToken(
  db: import('better-sqlite3').Database,
  config: { dbPath: string; secretKeyPath: string },
): Promise<string> {
  const settings = loadSettings(config)
  const clientId = (settings.google_client_id as string) || ''
  const clientSecret = (settings.google_client_secret as string) || ''
  const key = loadOrCreateKey(config.secretKeyPath)

  const token = await getValidAccessToken(db, key, clientId, clientSecret)
  if (!token) {
    throw new Error('Google not connected')
  }
  return token
}

export async function googleDataRoutes(app: FastifyInstance): Promise<void> {
  const config = (app as unknown as { config: { dbPath: string; secretKeyPath: string } }).config
  const db = (app as unknown as { db: import('better-sqlite3').Database }).db

  app.get('/api/google/calendars', {
    preHandler: [requireAuth],
  }, async (_request, reply) => {
    let accessToken: string
    try {
      accessToken = await getAccessToken(db, config)
    } catch {
      reply.code(400).send({ error: 'Google not connected' })
      return
    }
    return fetchCalendars(accessToken)
  })

  app.post('/api/google/photos/picker-session', {
    preHandler: [requireAuth],
  }, async (_request, reply) => {
    let accessToken: string
    try {
      accessToken = await getAccessToken(db, config)
    } catch {
      reply.code(400).send({ error: 'Google not connected' })
      return
    }
    const sessionData = await createPickerSession(accessToken)
    return {
      session_id: sessionData.id,
      picker_uri: sessionData.pickerUri,
      polling_config: sessionData.pollingConfig ?? {},
    }
  })

  app.get<{ Params: { id: string } }>('/api/google/photos/picker-session/:id', {
    preHandler: [requireAuth],
  }, async (request, reply) => {
    let accessToken: string
    try {
      accessToken = await getAccessToken(db, config)
    } catch {
      reply.code(400).send({ error: 'Google not connected' })
      return
    }
    const sessionData = await getPickerSession(accessToken, request.params.id)
    const mediaItemsSet = sessionData.mediaItemsSet
    const result: Record<string, unknown> = { media_items_set: mediaItemsSet }
    if (mediaItemsSet) {
      const items = await getSessionMediaItems(accessToken, request.params.id)
      result.photos = items.map((item) => ({
        id: item.id,
        url: `/api/photos/proxy?url=${encodeURIComponent(item.baseUrl)}`,
        mimeType: item.mimeType,
      }))
    }
    return result
  })

  app.delete<{ Params: { id: string } }>('/api/google/photos/picker-session/:id', {
    preHandler: [requireAuth],
  }, async (request, reply) => {
    let accessToken: string
    try {
      accessToken = await getAccessToken(db, config)
    } catch {
      reply.code(400).send({ error: 'Google not connected' })
      return
    }
    await deletePickerSession(accessToken, request.params.id)
    return { ok: true }
  })
}
